### PR TITLE
model : support Kimi-K3 recurrent-state rollback

### LINKED ISSUE #28461: model: support bounded recurrent-state rollback for Kimi-K3
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

## Requested behavior

Support bounded recurrent-state rollback for Kimi-K3 so speculative decoding can restore recurrent state at the accepted token boundary. Host checkpoints already provide a fallback; bounded snapshots could reduce that overhead.

At commit `74a7c897f049c17e7080423aa2111776eff6ebbf`, Kimi-K3 is excluded from `llm_arch_supports_rs_rollback`, and a requested nonzero `n_rs_seq` is clamped to zero. Its KDA layer stores only the final recurrent state and Q/K/V convolution windows.

Simply adding Kimi-K3 to the allowlist leaves historical snapshot groups unwritten. The diagnostic below exposes that gap. Stock Kimi-K3 has rollback disabled; the failing diagnostic requires both force-enabling rollback and initializing the recurrent cache with a nonzero byte pattern.

## Reproducer

Environment used for the recorded results: macOS, Apple clang, CPU-only Release build, and the generated dummy `kimi-k3-moe.gguf`. No real model weights are required.

Start from the pinned revision and apply two diagnostic changes: enable Kimi-K3 in the rollback allowlist, then add an environment guard for nonzero cache initialization. The second replacement changes only the constructor's buffer clear; the later clear path stays unchanged.

```sh
git clone https://github.com/ggml-org/llama.cpp.git llama-k3-rollback-repro
cd llama-k3-rollback-repro
git checkout 74a7c897f049c17e7080423aa2111776eff6ebbf

python3 - <<'PY'
from pathlib import Path

path = Path('src/llama-arch.cpp')
source = path.read_text()
old = 'bool llm_arch_supports_rs_rollback(const llm_arch & arch) {\n    switch (arch) {\n'
assert source.count(old) == 1
path.write_text(source.replace(old, old + '        case LLM_ARCH_KIMI_K3:\n', 1))

path = Path('src/llama-memory-recurrent.cpp')
source = path.read_text()
if '#include <cstdlib>' not in source:
    assert '#include <cstring>' in source
    source = source.replace('#include <cstring>', '#include <cstdlib>\n#include <cstring>', 1)
old = 'ggml_backend_buffer_clear(buf, 0);'
new = 'ggml_backend_buffer_clear(buf, std::getenv("RS_POISON") ? 0x3e : 0);'
assert old in source
path.write_text(source.replace(old, new, 1))
PY

cmake -B build-cpu -DGGML_METAL=OFF -DGGML_BLAS=OFF -DLLAMA_CURL=OFF \
      -DLLAMA_BUILD_TESTS=ON -DCMAKE_BUILD_TYPE=Release
cmake --build build-cpu -j6 --target test-llama-archs test-recurrent-state-rollback
mkdir -p models-generated
./build-cpu/bin/test-llama-archs -o models-generated/

env -u RS_POISON ./build-cpu/bin/test-recurrent-state-rollback \
    -m models-generated/kimi-k3-moe.gguf
RS_POISON=1 ./build-cpu/bin/test-recurrent-state-rollback \
    -m models-generated/kimi-k3-moe.gguf
```

`RS_POISON` is a temporary diagnostic guard added by the snippet, not an upstream option. Its presence fills the constructor's buffer with byte `0x3e`, producing the finite float32 pattern `0x3e3e3e3e` (about `0.18578431`). With the variable unset, initialization remains zero-filled.

## Recorded results

The existing `test_multi_seq_split_replay` uses two sequences, a 16-token microbatch, a 19-token prompt, rollback by three tokens, and a 40-token replay. Its comparison tolerance is `1e-7`.

| Kimi-K3 implementation | Zero-filled cache | Nonzero diagnostic cache |
| --- | --- | --- |
| Pinned revision plus allowlist entry only | Pass, max difference `3.66708e-09` | Fail, max difference `7.67533e-05`; first mismatch at sequence 0, position 16 |
| Same revision plus KDA and convolution snapshot stores | Pass, max difference `4.65661e-10` | Pass, max difference `4.65661e-10` |

Expected behavior is matching replay logits after restoring the accepted prefix. The failure appears when rollback reads a historical group that the original Kimi-K3 stores never populated. On the tiny dummy model, zero initialization hides the missing writes within the test tolerance.

The same nonzero-cache diagnostic passed for the already-supported generated Qwen3.5, Nemotron-H, and DeepSeek-V4 models, with multi-sequence replay differences of zero. These are recorded results from the earlier diagnostic; the temporary cache-initialization edit was subsequently reverted.

### Motivation

Support bounded recurrent-state rollback for Kimi-K3 so speculative decoding can restore recurrent state at the accepted token boundary. Host checkpoints already provide a fallback; bounded snapshots could reduce that overhead.

### Possible Implementation

Use the existing `build_recurrent_attn` helper for KDA state snapshots and store each Q/K/V convolution window at the corresponding rollback position. Enable the architecture in the rollback allowlist once both stores populate the snapshot groups. A local two-file patch implements this and produced the passing results above.

The stock dummy-model test alone does not detect the missing writes; improving that coverage remains a separate consideration. Full CI and other backends were not covered by this CPU diagnostic.

Related: [#28019](https://github.com/ggml-org/llama.cpp/issues/28019) reports rollback corruption for qwen4exp. Similar missing snapshot coverage is a possible connection, inferred from the two cases; this report concerns Kimi-K3.

AI usage disclosure: YES - AI assisted with implementation, diagnosis, and validation. Codex prepared this report from the source and recorded diagnostic results.

### PR DESCRIPTION
**Pull request by soulmachine**
## Overview

Enable bounded recurrent-state rollback for Kimi-K3 during speculative decoding. Kimi-K3 currently has rollback disabled and stores only the final KDA state and Q/K/V convolution windows. Enabling rollback without changing those stores would restore unwritten snapshot groups after rejecting draft tokens.

Save the convolution windows for each rollback position, use the existing `build_recurrent_attn` helper to save KDA state snapshots, and add Kimi-K3 to `llm_arch_supports_rs_rollback`.

Closes #28461.

Related: [#28019](https://github.com/ggml-org/llama.cpp/issues/28019) reports multi-sequence rollback corruption for qwen4exp. The connection to Kimi-K3's missing snapshots is an inference, not a maintainer-confirmed diagnosis. This change is limited to Kimi-K3 and does not resolve that issue.

## Additional information

Validation:

- `tests/CMakeLists.txt` registers the generated Kimi-K3 model with the existing rollback test. Each registered model runs both zero-filled and `0x3e`-filled cache passes in the test executable.
- CPU Release build on macOS with Apple clang passed. `GGML_SCHED_DEBUG_REALLOC=1 ctest --test-dir build-cpu -R 'recurrent-state-rollback' --output-on-failure`: 5/5 passed, including model generation and the four rollback tests.
- NVIDIA Vulkan Debug build on RTX PRO 6000 Blackwell passed the same 5/5 checks with `GGML_SCHED_NO_REALLOC=ON` and fatal warnings enabled.
- Both passes passed checkpoint restoration, multi-sequence split replay, and sequence-isolation checks for Qwen3.5, Nemotron-H, DeepSeek-V4, and Kimi-K3. Logit comparisons reject NaN and infinity; sequence-isolation checks matched exactly.

The nonzero pass uses the existing state-writing interface to collect and fill cache buffers entirely within the test. Buffer discovery decodes a full micro-batch to preserve Vulkan's prefill allocation sizes. Production cache initialization uses zeros; the `LLAMA_RS_DEBUG_FILL` hook has been removed. With only the Kimi-K3 allowlist change, the zero pass succeeded but the nonzero pass failed split replay (max logit difference `2.35e-6`, tolerance `1e-7`). Both passes succeed with the snapshot writes.

CI for `1ffc6ac3c`: 22 checks passed and four failed. The Kimi-K3 rollback test passes in all four failed jobs; the earlier Vulkan allocation abort is resolved.

| Failed job | Failure |
| --- | --- |
| [ROCm](https://github.com/ggml-org/llama.cpp/actions/runs/34106973977/job/101702142062) | Qwen3.5 split-replay mismatch in the zero-fill pass; also [seen on master](https://github.com/ggml-org/llama.cpp/actions/runs/34089517818/job/101639999492). |
| [Vulkan Apple](https://github.com/ggml-org/llama.cpp/actions/runs/34106973977/job/101702142174) | `test-thread-safety` bus error; also [seen on master](https://github.com/ggml-org/llama.cpp/actions/runs/34089517818/job/101639999950). |
| [Vulkan NVIDIA CM](https://github.com/ggml-org/llama.cpp/actions/runs/34106973977/job/101702141787) | `test-backend-ops` timed out after 1505 seconds on Tesla T4. |
| [Vulkan Intel Linux](https://github.com/ggml-org/llama.cpp/actions/runs/34106973977/job/101702142322) | `test-backend-ops` failed one `TOPK_MOE` numerical comparison. |

This PR does not change `test-backend-ops` or the GGML backends. The causes of the timeout and TOPK mismatch have not been confirmed.

Real-model measurements on September 5 used Kimi-K3 UD-IQ2_XXS on 8x RTX PRO 6000 Blackwell GPUs, with four server slots and seven draft tokens. The RadixArk DSpark draft shared GPU 7 with the target's output weights. These runs used a RelWithDebInfo build based on `4d91760`; DSpark also required a separate `t_layer_inp` registration patch, outside this PR. Throughput is in tok/s:

| Workload | No speculation | DSpark + host checkpoints | DSpark + rollback |
| --- | ---: | ---: | ---: |
| Rewrite | 12.88 | 17.7-19.6 | 22.19 |
| Knowledge | 12.81 | 11.9-12.4 | 15.75 |
| Code | 12.62 | 10.2-10.9 | 15.65 |
| Benchmark, concurrency 1 (1024 input / 256 output tokens) | 10.14 | Unavailable | 13.98 |
| Benchmark, concurrency 4 | 18.30 | Unavailable | 14.92 |

The rollback configuration was 13-25% faster on rewrite, 27-32% on knowledge, and 44-53% on code than the recorded host-checkpoint ranges. However, fit margins and model placement differed: the target's host model buffer was 5419.75 MiB with host checkpoints and 11010.12 MiB with rollback. These gains are not from a placement-controlled A/B.

Host-checkpoint results are unavailable for the concurrency benchmarks, so those rows cannot isolate rollback's contribution. Compared with no speculation, the combined DSpark and rollback configuration reduced TPOT from 77.7 to 49.2 ms at concurrency 1, but increased it from 134.5 to 220.6 ms at concurrency 4. These runs used random token inputs, disabled prompt caching, and four requests per concurrency slot.

Greedy outputs matched between DSpark with host checkpoints and DSpark with rollback on the three probe prompts. The rewrite output differed from the no-speculation run; knowledge and code outputs matched all three configurations. This does not establish output equivalence between DSpark and no speculation.

Snapshot storage multiplies recurrent-state memory by `1 + n_rs_seq`: the recorded four-slot cache grew from 1772.44 to 14179.50 MiB with seven rollback positions.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES - AI assisted with implementation, investigation, and validation. Codex drafted the commit message and this description from the patch and recorded results.

### CODE REVIEW

**Comment by CISC** [comment]
Thank you, reduce the comments though please.

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -238,6 +238,29 @@ if (NOT WIN32 OR NOT BUILD_SHARED_LIBS)
...
+    llama_test(
+        test-recurrent-state-rollback
+        NAME test-recurrent-state-rollback-kimi-k3-fill
+        LABEL main
+        ARGS -m "${MODEL_DIR}/kimi-k3-moe.gguf"
+    )
+    set_tests_properties(test-recurrent-state-rollback-kimi-k3-fill PROPERTIES
+        FIXTURES_REQUIRED generate-models
+        ENVIRONMENT "LLAMA_RS_DEBUG_FILL=0x3e"
+    )
```
Comment (CISC) [inline tests/CMakeLists.txt]: This should perhaps be a second pass in the test itself rather than a separate test so that all models get tested equally.

**Comment by soulmachine** [comment]
@CISC Shortened the code comments in eff6894ab, keeping the rollback invariant and test rationale.

File: tests/CMakeLists.txt (hunk 1 above)
Comment (soulmachine) [reply tests/CMakeLists.txt]: Moved this into a second pass of the test executable in 111ea913b and removed the separate CTest entry. All four models now run with both zero and 0x3e cache fills. The rollback suite passes (5/5).

**Review by CISC — APPROVED** [review]

File: src/llama-memory-recurrent.cpp [hunk 2]
```diff
@@ -114,13 +115,20 @@ llama_memory_recurrent::llama_memory_recurrent(
         }
     }
 
+    // nonzero initialization exposes missing snapshot writes in dummy-model tests.
+    const char * LLAMA_RS_DEBUG_FILL = getenv("LLAMA_RS_DEBUG_FILL");
+    const uint8_t rs_debug_fill = LLAMA_RS_DEBUG_FILL ? (uint8_t) strtoul(LLAMA_RS_DEBUG_FILL, nullptr, 0) : 0;
+    if (rs_debug_fill != 0) {
+        LLAMA_LOG_WARN("%s: filling the RS cache with 0x%02x instead of zeros (LLAMA_RS_DEBUG_FILL)\n", __func__, rs_debug_fill);
+    }
+
```
Comment (am17an) [inline src/llama-memory-recurrent.cpp]: remove this code

**Review by am17an — APPROVED** [review]

**Comment by soulmachine** [comment]
@ggerganov Could you review the latest fix and approve the pending CI runs when you have a chance? The Vulkan test failure is fixed in 1ffc6ac3c; targeted CPU and NVIDIA Vulkan checks pass (5/5 each). Thanks!

**Comment by soulmachine** [comment]
@ggerganov Could you rerun the [four failed CI jobs](https://github.com/ggml-org/llama.cpp/actions/runs/34106973977)? Kimi-K3 passes in all four. The ROCm Qwen rollback and Apple thread-safety failures also occur on master; the other two are a backend-test timeout and TOPK mismatch. Logs and comparisons are in the PR description. Thanks!

**Review by am17an — APPROVED** [review]

File: tests/test-recurrent-state-rollback.cpp [hunk 3]
```diff
@@ -1,22 +1,19 @@
 #include "arg.h"
 #include "common.h"
+#include "ggml-backend.h"
 #include "llama.h"
 
+#include "../src/llama-io.h"
+#include "../src/llama-memory.h"
```
Comment (ggerganov) [inline tests/test-recurrent-state-rollback.cpp:7]: Including internal headers should be avoided. At best, we can include `llama-ext.h`, but eventually it should also not be included. fyi @am17an @CISC

Comment (CISC) [reply tests/test-recurrent-state-rollback.cpp:7]: Any suggestions on how/what to move this to ext?

Comment (ggerganov) [reply tests/test-recurrent-state-rollback.cpp:7]: Not sure yet - I haven't looked into the logic of the new test and why it needs the headers. I would first look to write a test that does not require internal headers at all.

Comment (CISC) [reply tests/test-recurrent-state-rollback.cpp:7]: It didn't initially, but used an env-var instead.
